src/runtime-rs/crates/resource/src/cpu_mem/cpu.rs

```diff
@@ -0,0 +1,167 @@
+            .context("update container cpu resources")?;
+        let vcpu_required = self
+            .calc_cpu_resources()
+            .await
+            .context("calculate vcpus required")?;
+
+        if vcpu_required == self.current_vcpu().await {
+            return Ok(());
+        }
+
+        let curr_vcpus = self
+            .do_update_cpu_resources(vcpu_required, op, hypervisor, agent)
+            .await?;
+        self.update_current_vcpu(curr_vcpus).await;
+        Ok(())
+    }
+
+    async fn current_vcpu(&self) -> u32 {
+        let current_vcpu = self.current_vcpu.read().await;
+        *current_vcpu
+    }
+
+    async fn update_current_vcpu(&self, new_vcpus: u32) {
+        let mut current_vcpu = self.current_vcpu.write().await;
+        *current_vcpu = new_vcpus;
+    }
+
+    // update container_cpu_resources field
+    async fn update_container_cpu_resources(
+        &self,
+        cid: &str,
+        linux_cpus: Option<&LinuxCpu>,
+        op: ResourceUpdateOp,
+    ) -> Result<()> {
+        if let Some(cpu) = linux_cpus {
+            let container_resource = LinuxContainerCpuResources::try_from(cpu)?;
+            let mut resources = self.container_cpu_resources.write().await;
+            match op {
+                ResourceUpdateOp::Add | ResourceUpdateOp::Update => {
+                    resources.insert(cid.to_owned(), container_resource);
+                }
+                ResourceUpdateOp::Del => {
+                    resources.remove(cid);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    // calculates the total required vcpus by adding each container's requirements within the pod
+    async fn calc_cpu_resources(&self) -> Result<u32> {
+        let mut total_vcpu = 0;
+        let mut cpuset_vcpu: HashSet<u32> = HashSet::new();
+
+        let resources = self.container_cpu_resources.read().await;
+        for (_, cpu_resource) in resources.iter() {
+            let vcpu = cpu_resource.get_vcpus().unwrap_or(0) as u32;
+            cpuset_vcpu.extend(cpu_resource.cpuset().iter());
+            total_vcpu += vcpu;
+        }
+
+        // contrained only by cpuset
+        if total_vcpu == 0 && !cpuset_vcpu.is_empty() {
+            info!(sl!(), "(from cpuset)get vcpus # {:?}", cpuset_vcpu);
+            return Ok(cpuset_vcpu.len() as u32);
+        }
+
+        info!(
+            sl!(),
+            "(from cfs_quota&cfs_period)get vcpus count {}", total_vcpu
+        );
+        Ok(total_vcpu)
+    }
+
+    // do hotplug the vcpu
+    async fn do_update_cpu_resources(
+        &self,
+        new_vcpus: u32,
+        op: ResourceUpdateOp,
+        hypervisor: &dyn Hypervisor,
+        agent: &dyn Agent,
+    ) -> Result<u32> {
+        let old_vcpus = self.current_vcpu().await;
+
+        // when adding vcpus, ignore old_vcpus > new_vcpus
+        // when deleting vcpus, ignore old_vcpus < new_vcpus
+        if (op == ResourceUpdateOp::Add && old_vcpus > new_vcpus)
```

Comment:
Are there any inspections for the `ResourceUpdateOp::Update` option missing?

Reply:
We can not do this kind of check for ResourceUpdateOp, for we do not know whether the resources are required more or less.